src/math/exp_fcn.c

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+/* Include Files */
+#include <sof/math/exp_fcn.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/* Function Declarations */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b);
+
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *bitshiptrout,
+				uint64_t *bitsloptrout);
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	wide_mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = ((u64_rlo & BIT_MASK_LOW_Q27P5) != 0ULL);
+	u64_rlo =
+	    (u64_rhi << 36 | u64_rlo >> 28) + (roundup ? (uint64_t)1 : (uint64_t)0);
+	return (int64_t)u64_rlo;
+}
+
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *bitshiptrout
+ *		  uint64_t *bitsloptrout
+ * Return Type	: void
+ */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *bitshiptrout,
+				uint64_t *bitsloptrout)
+{
+	uint64_t absIn0;
+	uint64_t absIn1;
+	uint64_t in0Hi;
+	uint64_t in0Lo;
+	uint64_t in1Hi;
+	uint64_t prodHiLo;
+	uint64_t prodLoHi;
+
+	absIn0 = (in_0 < 0LL) ? ~(uint64_t)in_0 + 1ULL : (uint64_t)in_0;
+	absIn1 = (in_1 < 0LL) ? ~(uint64_t)in_1 + 1ULL : (uint64_t)in_1;
+	in0Hi = absIn0 >> 32ULL;
+	in0Lo = absIn0 & UINT32_MAX;
+	in1Hi = absIn1 >> 32ULL;
+	absIn0 = absIn1 & UINT32_MAX;
+	prodHiLo = in0Hi * absIn0;
+	prodLoHi = in0Lo * in1Hi;
+	absIn0 *= in0Lo;
+	absIn1 = 0ULL;
+	in0Lo = absIn0 + (prodLoHi << 32ULL);
+	if (in0Lo < absIn0)
+		absIn1 = 1ULL;
+
+	absIn0 = in0Lo;
+	in0Lo += prodHiLo << 32ULL;
+	if (in0Lo < absIn0)
+		absIn1++;
+
+	absIn0 = ((absIn1 + in0Hi * in1Hi) + (prodLoHi >> 32ULL)) + (prodHiLo >> 32ULL);
+	if (in_0 != 0LL && (in_1 != 0LL && ((in_0 > 0LL) != (in_1 > 0LL)))) {
```

Comment:
wow... really way too many

Reply:
During CI inspections, I notice mistakes 

```
/home/runner/work/sof/sof/src/math/exp_fcn.c:75:43: error: suggest parentheses around comparison in operand of ‘!=’ [-Werror=parentheses]
   75 |  if (in_0 != 0LL && (in_1 != 0LL && (in_0 > 0LL != in_1 > 0LL))) {
      |                                      ~~~~~^~~~~
      
/home/runner/work/sof/sof/src/math/exp_fcn.c:75:59: error: suggest parentheses around comparison in operand of ‘!=’ [-Werror=parentheses]
   75 |  if (in_0 != 0LL && (in_1 != 0LL && ((in_0 > 0LL) != in_1 > 0LL))) {
      |                                                      ~~~~~^~~~~      
```

Reply:
```
if (in_0 != 0LL && in_1 != 0LL && (in_0 > 0LL) != (in_1 > 0LL))
```
has a lot fewer parentheses, does the same and shouldn't generate any warnings

Reply:
I'm afraid the advice given is still insufficient.  The parenthesis continue to fail the CI strict check. 

For  reference I am adding output
```
CHECK: Unnecessary parentheses around 'in_0 > 0LL'
#[17](https://github.com/thesofproject/sof/actions/runs/3549190428/jobs/5961273777#step:5:18)1: FILE: src/math/exp_fcn.c:75:
+	if (in_0 != 0LL && in_1 != 0LL && (in_0 > 0LL) != (in_1 > 0LL)) {

CHECK: Unnecessary parentheses around 'in_1 > 0LL'
#171: FILE: src/math/exp_fcn.c:75:
+	if (in_0 != 0LL && in_1 != 0LL && (in_0 > 0LL) != (in_1 > 0LL)) {
```

Reply:
there you go. I was even too conservative. You can drop those last internal parentheses too then

Reply:
still, a warning is present!! 
```
/home/runner/work/sof/sof/src/math/exp_fcn.c:80:37: error: suggest parentheses around comparison in operand of ‘!=’ [-Werror=parentheses]
   80 |  if (in_0 != 0 && in_1 != 0 && in_0 > 0 != in_1 > 0) {
      |                                ~~~~~^~~
```

Reply:
> The parenthesis continue to fail the CI strict check.

This is not a "CI" check, `-Werror=parentheses` is a compiler check. As you failed to share the link to the logs where this appeared, I'm guessing it's the Zephyr SDK. Please install it and try it, it's just one big tarball to download and extra in your home directory. Then you temporarily undefine XTENSA_TOOLS_ROOT and that's it.

Reply:
> is it just me or are the two warnings above mutually contradictory?

It's hard to tell because this code has kept changing and the links to the logs have been lost but I think one warnings is from checkpatch and the other is from the  compiler in the Zephyr toolchain.

A compiler almost always wins over checkpatch for many reasons:
- A compiler understands the code much better
- A compiler is required, checkpatch is not
- "Checkpatch is not always right" https://www.kernel.org/doc/html/latest/dev-tools/checkpatch.html
- Last but not least, a compiler warning would _stick after this PR is merged_!! A checkpatch warning will not.

Reply:
> "Checkpatch is not always right" https://www.kernel.org/doc/html/latest/dev-tools/checkpatch.html

checkpatch is more often wrong when using `checkpatch --strict` a.k.a. `checkpatch --subjective`. 

In https://github.com/thesofproject/sof/actions/runs/3564456798/jobs/5988433936, only `--subjective` is complaining about parentheses. So we're good.

Reply:
> checkpatch is more often wrong when using checkpatch --strict a.k.a. checkpatch --subjective.

I'm renaming `--strict` to `--subjective` in https://github.com/thesofproject/sof/pull/6689 to make that more obvious. One-line change, please review.

Reply:
> is it just me or are the two warnings above mutually contradictory? One is suggesting parentheses and another one recommends to remove them. Looks like we cannot have both

It doesn't work either ,
```
/home/runner/work/sof/sof/src/math/exp_fcn.c:92:44: error: suggest parentheses around comparison in operand of ‘!=’ [-Werror=parentheses]
   92 |         if (in_0 != 0 && in_1 != 0 && in_0 > 0 != in_1 > 0) {
      |                                       ~~~~~^~~

```      
Do you have a more effective solution?
I would have to put up with warnings that were below code otherwise.
`if (in_0 != 0 && in_1 != 0 && in_0 > 0 != (in_1 > 0)) {`